# üìò Aula - Try/Catch

## Introdu√ß√£o

Como j√° foi mencionado em aulas anteriores, o modelo de fun√ß√£o async √© mais adequado para realiza√ß√£o de requisi√ß√µes porque tem uma leitura melhor e porque √© mais escal√°vel. Permitindo evitar o famoso: promise callback hell.

Agora, para deixarmos esse uso ainda mais refinado, vamos aprender sobre a estrutura trycatch.

## O trycatch

O trycatch consiste numa estrutura de c√≥digo nativa do Javascript que permite lidarmos com erros de forma inteligente, muito semelhante ao then.catch, mas adequado para o modelo de fun√ß√£o ass√≠ncrona que estamos trabalhando.

```js
try{
‚Å†
} catch (error) {
‚Å†
}
```

Dentro de try, o Javascript vai tentar executar cada linha de c√≥digo, e caso haja um erro, ele vai interromper a execu√ß√£o e executar a instru√ß√£o que est√° presente no catch.

Confira um exemplo contextualizado:

```js
async function loadData() {
   try {
      const response = await api.get("/fruits", {
         params: {
            category: category,
         },
      });            
      setFruitList(response.data);
   } catch (error) {
      console.log(error);
   } 
}      
```
Nesse exemplo, temos em try, uma tentativa de requisi√ß√£o seguida de uma atualiza√ß√£o de estado, caso a requisi√ß√£o d√™ certo a atualiza√ß√£o vai se concretizar, do contr√°rio cairemos em catch, reproduzindo o erro no console.

## Finally

Podemos aprimorar ainda mais esta estrutura, adicionando o bloco "finally" √† estrutura "try-catch". O bloco "finally" √© utilizado quando desejamos executar uma instru√ß√£o ap√≥s a tentativa, **independentemente de ter sido bem-sucedida ou ter ocorrido um erro**.

Confira o exemplo contextualizado:
```js
async function loadData() {
   try {
      setLoading(true);
      const response = await api.get("/fruits", {
         params: {
            category: category,
         },
      });            
      setFruitList(response.data);
   } catch (error) {
      console.log(error);
   } finally {
      setLoading(false);
   }‚Å†

}   
```

Aprimoramos nossa funcionalidade adicionando um comportamento de carregamento. Agora, ao iniciarmos a tentativa, definimos nosso status de carregamento como verdadeiro. E, independentemente do resultado da tentativa, ao t√©rmino, no bloco finally, suspendemos nosso carregamento.

## Conclus√£o

Nesta aula, aprendemos sobre o trycatch e como podemos lidar com requisi√ß√µes com mais qualidade. 
